mlflow/langchain/_langchain_autolog.py

```diff
@@ -238,40 +249,121 @@ def _update_langchain_model_config(model):
+        isinstance(new_callback, type(original_callback))
+        for original_callback in original_callbacks
+    )
+
+
+def _is_base_callback_manager(callback):
     from langchain_core.callbacks.base import BaseCallbackManager
 
-    def _is_new_callback_already_in_original_callbacks(new_callback, original_callbacks):
-        return any(
-            isinstance(new_callback, type(original_callback))
-            for original_callback in original_callbacks
-        )
+    return isinstance(callback, BaseCallbackManager)
 
-    if isinstance(original_callbacks, BaseCallbackManager):
-        for callback in new_callbacks:
-            if not _is_new_callback_already_in_original_callbacks(
-                callback, original_callbacks.handlers
-            ):
-                original_callbacks.add_handler(callback)
-        return original_callbacks
 
-    if not isinstance(original_callbacks, list):
-        original_callbacks = [original_callbacks]
+def _inject_callbacks_into_base_callback_manager(original_callbacks, new_callbacks):
+    added_callbacks = []
+    for callback in new_callbacks:
+        if not _is_new_callback_already_in_original_callbacks(
+            callback, original_callbacks.handlers
+        ):
+            added_callbacks.append(callback)
+            original_callbacks.add_handler(callback)
+    return original_callbacks, added_callbacks
+
+
+def _is_list_of_base_callback_handlers(callback):
+    from langchain_core.callbacks.base import BaseCallbackHandler
+
+    return isinstance(callback, list) and all(
+        isinstance(callback_handler, BaseCallbackHandler) for callback_handler in callback
+    )
+
 
+def _inject_callbacks_into_list_of_base_callback_handlers(original_callbacks, new_callbacks):
     # Make a copy of the original callbacks to avoid modifying the original list
     updated_callbacks = list(original_callbacks)
     for new_callback in new_callbacks:
         if not _is_new_callback_already_in_original_callbacks(new_callback, original_callbacks):
             updated_callbacks.append(new_callback)
-
     return updated_callbacks
 
 
+@contextlib.contextmanager
+def _inject_callbacks(original_callbacks, new_callbacks):
+    """
+    Inject list of callbacks into the original callbacks.
+    For RunnableConfig, the callbacks is defined as List[BaseCallbackHandler] or BaseCallbackManager
+    https://github.com/langchain-ai/langchain/blob/ed980601e1c630f996aabf85df5cb26178e53099/libs/core/langchain_core/callbacks/base.py#L636
+
+    Args:
+        original_callbacks: List[BaseCallbackHandler] or BaseCallbackManager
+        new_callbacks: List[BaseCallbackHandler]
+    """
+    if _is_base_callback_manager(original_callbacks):
+        added_callbacks = []
+        try:
+            original_callbacks, added_callbacks = _inject_callbacks_into_base_callback_manager(
+                original_callbacks, new_callbacks
+            )
+            yield original_callbacks
+        except Exception as e:
```

Comment:
Is there any particular reason we catch exception only here?

Reply:
Ah at first I thought I only need to change here as the list is not mutated, so added try catch 🤣 Seems we don't need to catch it as autologging catch all exceptions